File: Tests/ObjC/VSS007_HighLevelTests.m
//
//  VSS007_HighLevelTests.m
//  VirgilSDK
//
//  Created by Oleksandr Deundiak on 11/9/16.
//  Copyright © 2016 VirgilSecurity. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <XCTest/XCTest.h>

#import "VSSTestsUtils.h"

/// Each request should be done less than or equal this number of seconds.
static const NSTimeInterval kEstimatedRequestCompletionTime = 5.;

@interface VSS007_HighLevelTests : XCTestCase

//@property (nonatomic) VSSTestsConst *consts;
//@property (nonatomic) VSSTestsUtils *utils;
//@property (nonatomic) VSSVirgilKey *key;
//@property (nonatomic) VSSVirgilAuthorityKey *authorityKey;

@end

@implementation VSS007_HighLevelTests

#pragma mark - Setup

//- (void)setUp {
//    [super setUp];
//    
//    self.consts = [[VSSTestsConst alloc] init];
//    self.utils = [[VSSTestsUtils alloc] initWithCrypto:[[VSSCrypto alloc] init] consts:self.consts];
//    
//    [VSSVirgilConfig initializeWithApplicationToken:self.consts.applicationToken];
//    
//    NSData *appPrivateKey = [[NSData alloc] initWithBase64EncodedString:self.consts.applicationPrivateKeyBase64 options:0];
//    self.authorityKey = [VSSVirgilAuthorityKey virgilAuthorityKeyWithData:appPrivateKey password:self.consts.applicationPrivateKeyPassword forAppId:self.consts.applicationId];
//}
//
//- (void)tearDown {
//    self.consts = nil;
//    self.authorityKey = nil;
//    self.key = nil;
//    
//    [super tearDown];
//}
//
//- (void)test001_CreateCard {
//    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Virgil Card should be created."];
//    
//    NSUInteger numberOfRequests = 1;
//    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
//    
//    NSString *uuidStr = [[NSUUID UUID] UUIDString];
//    
//    self.key = [VSSVirgilKey virgilKeyWithName:uuidStr];
//    
//    XCTAssert(self.key != nil);
//    
//    NSError *error;
//    VSSCreateCardRequest *request = [self.key buildCreateCardRequestWithIdentity:uuidStr identityType:self.consts.applicationIdentityType data:nil error:&error];
//    XCTAssert(request != nil);
//    XCTAssert(error == nil);
//    
//    [self.authorityKey signRequest:request error:&error];
//    XCTAssert(error == nil);
//    
//    [VSSVirgilCard createCardWithRequest:request completion:^(VSSVirgilCard *card, NSError *error) {
//        XCTAssert(card != nil);
//        XCTAssert(error == nil);
//        
//        XCTAssert([card.identifier length] > 0);
//        XCTAssert([self.utils checkVirgilCard:card isEqualToCreateCardRequest:request]);
//        
//        [ex fulfill];
//    }];
//    
//    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
//        XCTAssert(error == nil);
//    }];
//}
////
//- (void)test002_CreateCardWithData {
//    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Virgil Card with data should be created."];
//    
//    NSUInteger numberOfRequests = 1;
//    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
//    
//    VSSCreateCardRequest *request = [self.utils instantiateCreateCardRequestWithData:@{
//                                                                                       @"customKey1" : @"customField1",
//                                                                                       @"customKey2" : @"customField2"}];
//    
//    [self.client createCardWithRequest:request completion:^(VSSCard *card, NSError *error) {
//        if (error != nil) {
//            XCTFail(@"Expectation failed: %@", error);
//            return;
//        }
//        
//        XCTAssert([card.identifier length] > 0);
//        XCTAssert([self.utils checkCard:card isEqualToCreateCardRequest:request]);
//        
//        [ex fulfill];
//    }];
//    
//    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
//        if (error != nil)
//            XCTFail(@"Expectation failed: %@", error);
//    }];
//}
//
//- (void)test003_SearchCards {
//    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Virgil Card should be created. Search should return 1 card which is equal to created card"];
//    
//    NSUInteger numberOfRequests = 2;
//    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
//    
//    VSSCreateCardRequest *request = [self.utils instantiateCreateCardRequest];
//    
//    [self.client createCardWithRequest:request completion:^(VSSCard *card, NSError *error) {
//        if (error != nil) {
//            XCTFail(@"Expectation failed: %@", error);
//            return;
//        }
//        
//        VSSSearchCardsCriteria *searchCardsCriteria = [VSSSearchCardsCriteria searchCardsCriteriaWithScope:VSSCardScopeApplication identityType:card.identityType identities:@[card.identity]];
//        
//        sleep(3);
//        [self.client searchCardsUsingCriteria:searchCardsCriteria completion:^(NSArray<VSSCard *>* cards, NSError *error) {
//            if (error != nil) {
//                XCTFail(@"Expectation failed: %@", error);
//                return;
//            }
//            
//            XCTAssert([cards count] == 1);
//            XCTAssert([self.utils checkCard:cards[0] isEqualToCard:card]);
//            
//            [ex fulfill];
//        }];
//    }];
//    
//    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
//        if (error != nil)
//            XCTFail(@"Expectation failed: %@", error);
//    }];
//}
//
//- (void)test004_GetCard {
//    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Virgil Card should be created. Get card request should return 1 card which is equal to created card"];
//    
//    NSUInteger numberOfRequests = 2;
//    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
//    
//    VSSCreateCardRequest *request = [self.utils instantiateCreateCardRequest];
//    
//    [self.client createCardWithRequest:request completion:^(VSSCard *card, NSError *error) {
//        if (error != nil) {
//            XCTFail(@"Expectation failed: %@", error);
//            return;
//        }
//        
//        sleep(3);
//        [self.client getCardWithId:card.identifier completion:^(VSSCard *foundCard, NSError *error) {
//            if (error != nil) {
//                XCTFail(@"Expectation failed: %@", error);
//                return;
//            }
//            
//            XCTAssert(foundCard != nil);
//            XCTAssert([foundCard.identifier isEqualToString:card.identifier]);
//            XCTAssert([self.utils checkCard:foundCard isEqualToCreateCardRequest:request]);
//            
//            [ex fulfill];
//        }];
//    }];
//    
//    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
//        if (error != nil)
//            XCTFail(@"Expectation failed: %@", error);
//    }];
//}
//
//- (void)test005_RevokeCard {
//    XCTestExpectation * __weak ex = [self expectationWithDescription:@"Virgil Card should be created. Virgil card should be revoked"];
//    
//    NSUInteger numberOfRequests = 3;
//    NSTimeInterval timeout = numberOfRequests * kEstimatedRequestCompletionTime;
//    
//    VSSCreateCardRequest *request = [self.utils instantiateCreateCardRequest];
//    
//    [self.client createCardWithRequest:request completion:^(VSSCard *card, NSError *error) {
//        if (error != nil) {
//            XCTFail(@"Expectation failed: %@", error);
//            return;
//        }
//        
//        VSSRevokeCardRequest *revokeRequest = [self.utils instantiateRevokeCardForCard:card];
//        
//        sleep(3);
//        [self.client revokeCardWithRequest:revokeRequest completion:^(NSError *error) {
//            if (error != nil) {
//                XCTFail(@"Expectation failed: %@", error);
//                return;
//            }
//            
//            [self.client getCardWithId:card.identifier completion:^(VSSCard *foundCard, NSError *error) {
//                XCTAssert(foundCard == nil);
//                
//                XCTAssert(error != nil);
//                XCTAssert(error.code == 404);
//                
//                [ex fulfill];
//            }];
//        }];
//    }];
//    
//    [self waitForExpectationsWithTimeout:timeout handler:^(NSError *error) {
//        if (error != nil)
//            XCTFail(@"Expectation failed: %@", error);
//    }];
//}

@end

File: Tests/Swift/VSS007_HighLevelTests.swift
//
//  VSS007_HighLevelTests.swift
//  VirgilSDK
//
//  Created by Oleksandr Deundiak on 3/3/17.
//  Copyright © 2017 VirgilSecurity. All rights reserved.
//

import Foundation
import XCTest
import VirgilSDK

class VSS007_HighLevelTests: XCTestCase {
    private var utils: VSSTestUtils!
    private var consts: VSSTestsConst!
    private var api: VSSVirgilApi!
    private var mailinator: Mailinator!
    
    // MARK: Setup
    override func setUp() {
        super.setUp()
        
        self.consts = VSSTestsConst()
        
        let appPrivateKeyData = Data(base64Encoded: self.consts.applicationPrivateKeyBase64)!
        let credentials = VSSCredentials(appKeyData: appPrivateKeyData, appKeyPassword: self.consts.applicationPrivateKeyPassword, appId: self.consts.applicationId)
        
        let crypto = VSSCrypto()
        
        let context = VSSVirgilApiContext(crypto: VSSCrypto(), token: self.consts.applicationToken, credentials: credentials, cardVerifiers: nil)
        let validator = VSSCardValidator(crypto: crypto)
        let privateKey = crypto.importPrivateKey(from: Data(base64Encoded: self.consts.applicationPrivateKeyBase64)!, withPassword: self.consts.applicationPrivateKeyPassword)!
        let publicKey = crypto.extractPublicKey(from: privateKey)
        let publicKeyData = crypto.export(publicKey)
        XCTAssert(validator.addVerifier(withId: self.consts.applicationId, publicKeyData: publicKeyData))
        // FIXME
        validator.useVirgilServiceVerifiers = false
        let config = VSSServiceConfig(token: self.consts.applicationToken)
        config.cardValidator = validator
        
//        config.cardsServiceURL = self.consts.cardsServiceURL
//        config.cardsServiceROURL = self.consts.cardsServiceROURL
//        config.identityServiceURL = self.consts.identityServiceURL
//        config.registrationAuthorityURL = self.consts.registrationAuthorityURL
        
        let client = VSSClient(serviceConfig: config)
        context.client = client
        
        self.api = VSSVirgilApi(context: context)
        
        self.utils = VSSTestUtils(crypto: VSSCrypto(), consts: self.consts)
        
        self.mailinator = Mailinator(applicationToken: self.consts.mailinatorToken, serviceUrl: URL(string: "https://api.mailinator.com/api/")!)
    }
    
    override func tearDown() {
        self.utils = nil
        self.consts = nil
        self.api = nil
        
        super.tearDown()
    }
    
    // MARK: Tests
    func testC01_CreateCard() {
        let ex = self.expectation(description: "Virgil Card should be created")
        
        let numberOfRequests = 1
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime)
        
        let identityValue = NSUUID().uuidString
        let identity = self.api.identities.createUserIdentity(withValue: identityValue, type: "test")
        let key = self.api.keys.generateKey()
        let card = try! self.api.cards.createCard(with: identity, ownerKey: key)
        self.api.cards.publish(card) { (error) in
            XCTAssert(error == nil)
            XCTAssert(self.utils.checkHighLevelCardIsValid(card: card))
            
            ex.fulfill()
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testC02_CreateCardWithData() {
        let ex = self.expectation(description: "Virgil Card with data should be created")
        
        let numberOfRequests = 1
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime)
        
        let identityValue = NSUUID().uuidString
        let identity = self.api.identities.createUserIdentity(withValue: identityValue, type: "test")
        let key = self.api.keys.generateKey()
        
        let card = try! self.api.cards.createCard(with: identity, ownerKey: key, customFields: ["custom_key1": "custom_field1", "custom_key2": "custom_field2"])
        self.api.cards.publish(card) { (error) in
            XCTAssert(error == nil)
            XCTAssert(self.utils.checkHighLevelCardIsValid(card: card))
            
            ex.fulfill()
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testC03_SearchCards() {
        let ex = self.expectation(description: "Virgil Card should be created. Search should return 1 card which is equal to created card");
        
        let numberOfRequests = 2
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime)
        
        let identityValue = NSUUID().uuidString
        let identity = self.api.identities.createUserIdentity(withValue: identityValue, type: "test")
        let key = self.api.keys.generateKey()
        
        let card = try! self.api.cards.createCard(with: identity, ownerKey: key)
        self.api.cards.publish(card) { (error) in
            sleep(3)
            self.api.cards.searchCards(withIdentities: [identityValue]) { cards, error in
                XCTAssert(error == nil)
                XCTAssert(cards != nil)
                XCTAssert(cards!.count == 1)
                
                XCTAssert(self.utils.check(card: card, isEqualToCard: cards![0]))
                
                ex.fulfill()
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testC04_GetCard() {
        let ex = self.expectation(description: "Virgil Card should be created. Get card request should return 1 card which is equal to created card");
        
        let numberOfRequests = 2
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime)
        
        let identityValue = NSUUID().uuidString
        let identity = self.api.identities.createUserIdentity(withValue: identityValue, type: "test")
        let key = self.api.keys.generateKey()
        
        let card = try! self.api.cards.createCard(with: identity, ownerKey: key)
        self.api.cards.publish(card) { (error) in
            sleep(3)
            
            self.api.cards.getCard(withId: card.identifier) { foundCard, error in
                XCTAssert(error == nil)
                XCTAssert(foundCard != nil)
                
                XCTAssert(self.utils.check(card: card, isEqualToCard: foundCard!))
                
                ex.fulfill()
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testC05_RevokeCard() {
        let ex = self.expectation(description: "Virgil Card should be created. Virgil card should be revoked");
        
        let numberOfRequests = 3
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime)
        
        let identityValue = NSUUID().uuidString
        let identity = self.api.identities.createUserIdentity(withValue: identityValue, type: "test")
        let key = self.api.keys.generateKey()
        
        let card = try! self.api.cards.createCard(with: identity, ownerKey: key)
        self.api.cards.publish(card) { (error) in
            sleep(3)
            
            self.api.cards.revoke(card) { error in
                XCTAssert(error == nil)
                ex.fulfill()
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testA01_CreateGlobalEmailCard() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created")
        
        let numberOfRequests = 5
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + kEstimatedEmailReceiveTime)
        
        let identityValue = self.utils.generateEmail()
        let identity = self.api.identities.createEmailIdentity(withEmail: identityValue)
        
        identity.check(options: nil) { error in
            self.utils.getConfirmationCode(identityValue: identityValue, mailinator: self.mailinator) { confirmationCode in
                identity.confirm(withConfirmationCode: confirmationCode) { error in
                    let key = self.api.keys.generateKey()
                    
                    let card = try! self.api.cards.createCard(with: identity, ownerKey: key)
                    self.api.cards.publish(card) { (error) in
                        XCTAssert(error == nil)
                        XCTAssert(self.utils.checkHighLevelCardIsValid(card: card))
                        
                        ex.fulfill()
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
    
    func testA02_RevokeGlobalEmailCard() {
        let ex = self.expectation(description: "Global Email Virgil Card should be created. Global Email Virgil Card should be revoked.")
        
        let numberOfRequests = 5
        let timeout = TimeInterval(numberOfRequests * kEstimatedRequestCompletionTime + 2 * kEstimatedEmailReceiveTime)
        
        let identityValue = self.utils.generateEmail()
        let identity = self.api.identities.createEmailIdentity(withEmail: identityValue)
        
        identity.check(options: nil) { error in
            self.utils.getConfirmationCode(identityValue: identityValue, mailinator: self.mailinator) { confirmationCode in
                identity.confirm(withConfirmationCode: confirmationCode) { error in
                    let key = self.api.keys.generateKey()
                    
                    let card = try! self.api.cards.createCard(with: identity, ownerKey: key)
                    self.api.cards.publish(card) { (error) in
                        let identity = self.api.identities.createEmailIdentity(withEmail: identityValue)
                        
                        identity.check(options: nil) { error in
                            self.utils.getConfirmationCode(emailNumber: 1, identityValue: identityValue, mailinator: self.mailinator) { confirmationCode in
                                identity.confirm(withConfirmationCode: confirmationCode) { error in
                                    self.api.cards.revokeEmail(card, identity: identity, ownerKey: key) { error in
                                        XCTAssert(error == nil)
                                        
                                        ex.fulfill()
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        
        self.waitForExpectations(timeout: timeout) { error in
            guard error == nil else {
                XCTFail("Expectation failed: " + error!.localizedDescription)
                return
            }
        }
    }
}
